### ISSUE TITLE
Improving the forward pass

### ISSUE THREAD

**Issue by philipturner** [Dao-AILab/flash-attention#308]
IIUC the backward pass is difficult to parallelize, as there are data dependencies going toward both $dQ$ and $dK$. Thus a single GPU threadgroup must tackle an entire attention matrix. To fully utilize the GPU, you need `batch * heads` to vastly exceed the number of GPU cores, which might not always be true.

The design of the backward pass seemed to leak into the forward pass, making it less parallelizable and quite convoluted. In the paper, it repeatedly scales/descales blocks of O by the softmax normalization factor. This reduces numerical accuracy and increases compute cost. A simpler approach would parallelize a single NxN attention matrix row-wise. Hold Q/O in registers (O in FP32) and fetch K/V from HBM each iteration. Don't scale O by $l^{-1}$ until the very end of the GPU kernel.

**Comment by tridao** [comment]
Wonderful! This version is actually already implemented and will be released soon.

**Comment by tridao** [comment]
To clarify, the Triton version in this [repo](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py) already implements things as you described :D. 
There's an updated CUDA implementation that also does things this way and will be released soon.

**Comment by philipturner** [comment]
You can also elide computations falling over irregular/non-implicit attention masks, by computing a block-sparse mask dynamically at runtime. Then give the mask 3 flags:

- (0) Mask is entirely -MAXHALF; skip this loop iteration
- (1) Mask is entirely zero; skip the mask loading
- (2) Mask is mixed; perform the iteration with mask loading

This should help with language models, where the mask is approximately triangular.
- Skips unnecessary computations
- Minimizes HBM accesses to the mask
- Minimizes costly reads of unaligned matrix edges when `seq_len % block size != 0`

**Comment by tridao** [comment]
This is great! More elegant than the approach I was taking, which is to specialize an implementation for attention with `causal=True` (to skip about half the computation), and another specialized implementation with a block sparse mask.

**Comment by philipturner** [comment]
I was thinking of only supporting implicit, perfectly triangular masks, but real-world code performs some non-deterministic mutations to the attention matrix.

Another improvement would perform load balancing across the unequal upper/lower rows of the triangular matrix. Dispatch threadgroups in reverse order along rows `GROUP_ID = seq_len -> seq_len / 2`. Once each threadgroup finishes its first row, it jumps to row `seq_len - GROUP_ID` but traverses columns in backwards order. This maximizes cache coherence (w.r.t. other groups) once the group starts hitting non-masked blocks again.
